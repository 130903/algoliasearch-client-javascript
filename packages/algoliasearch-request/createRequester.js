// @flow

import RequestHosts from './hosts';

import type {
  AppId,
  ApiKey,
  RequestOptions,
  RequestArguments,
} from '../algoliasearch/src/types';
import type { HttpModule, Timeouts, Hosts } from './types';

type Args = {|
  appId: AppId,
  apiKey: ApiKey,
  httpRequester: HttpModule,
  options?: {|
    timeouts?: Timeouts,
    extraHosts?: Hosts,
  |},
  requestOptions?: RequestOptions,
|};

const stringify = qs => JSON.stringify(qs); // todo: use proper url stringify

type ErrorTypes = 'application' | 'network' | 'dns' | 'timeout';
const retryableErrors: Array<ErrorTypes> = [
  'application',
  'network',
  'dns',
  'timeout',
];

export class Requester {
  hosts: RequestHosts;
  apiKey: ApiKey;
  appId: AppId;
  requestOptions: RequestOptions;
  requester: HttpModule;

  constructor({
    appId,
    apiKey,
    httpRequester,
    options = {},
    requestOptions = {},
  }: Args) {
    if (typeof appId !== 'string') {
      throw new Error(
        `appId is required and should be a string, received ${appId}`
      );
    }
    if (typeof apiKey !== 'string') {
      throw new Error(
        `apiKey is required and should be a string, received ${apiKey}`
      );
    }
    if (typeof httpRequester !== 'function') {
      throw new Error(
        `httpRequester is required and should be a function, received ${httpRequester}`
      );
    }
    this.hosts = new RequestHosts({ appId, ...options });
    this.appId = appId;
    this.apiKey = apiKey;
    this.requester = httpRequester;
    this.requestOptions = requestOptions;
  }

  setOptions = (fn: RequestOptions => RequestOptions): RequestOptions => {
    const oldOptions = this.requestOptions;
    const newOptions = fn(oldOptions);
    this.requestOptions = newOptions;
    return newOptions;
  };

  request = ({
    method,
    path,
    qs,
    body,
    options,
    requestType: type,
    retry = 0,
  }: RequestArguments) => {
    const hostname = this.hosts.getHost({ type });
    const timeout = this.hosts.getTimeout({ retry, type });

    const pathname = path + stringify(qs);
    const url = { hostname, pathname };

    return new Promise((resolve, reject) => {
      this.requester({
        body,
        method,
        url,
        timeout,
        options,
      })
        .then(resolve)
        .catch(err => {
          if (retryableErrors.indexOf(err.message.reason) > -1) {
            resolve(
              this.request({
                method,
                path,
                qs,
                body,
                options,
                requestType: type,
                retry: retry + 1,
              })
            );
          }
          reject(err);
        });
    });
  };
}

// todo: not use this and just use the class
export default function createRequester(args: Args) {
  const _r = new Requester(args);
  const requester = _r.request;
  requester.setOptions = _r.setOptions;
  requester.options = _r.requestOptions;
  return requester;
}
