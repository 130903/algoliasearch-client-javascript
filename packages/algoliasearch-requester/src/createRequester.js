// @flow
import { createMemoryStore } from 'universal-store';
import { AlgoliaError, AlgoliaRequesterError } from 'algoliasearch-errors';
import {
  initHostAndTimeouts,
  getParams,
  hostDidFail,
  hostDidTimeout,
} from './HostAndTimeoutManager.js';

import type { AppId, ApiKey } from 'algoliasearch';
import type { Store, Data } from 'universal-store';
import type {
  RequestOptions,
  RequesterOptions,
  RequestArguments,
  Result,
  HttpModule,
  ErrorType,
  CreateRequester,
} from 'algoliasearch-requester';

const stringify = qs => JSON.stringify(qs); // todo: use proper url stringify

const toCacheKey = obj => JSON.stringify(obj); // todo: find if something is faster

const retryableErrors: Array<ErrorType> = ['server', 'network', 'timeout'];

// eslint-disable-next-line no-unused-vars
const RESET_HOST_TIMER = 120000; // ms; 2 minutes
// eslint-disable-next-line no-unused-vars
const RESET_TIMEOUT_TIMER = 1200000; // ms; 20 minutes

export class Requester {
  apiKey: ApiKey;
  appId: AppId;
  requestOptions: RequestOptions;
  requester: HttpModule;
  cache: boolean;
  store: Store<string, Data>;
  requesterOptions: RequesterOptions;

  constructor({
    appId,
    apiKey,
    httpRequester,
    requestOptions,
    // todo: put timeouts in requestOptions
    requesterOptions: { timeouts = {}, cache = false, hosts } = {},
  }: {|
    appId: AppId,
    apiKey: ApiKey,
    httpRequester: HttpModule,
    requestOptions?: RequestOptions,
    requesterOptions?: RequesterOptions,
  |}) {
    if (typeof appId !== 'string') {
      throw new AlgoliaError(
        `appId is required and should be a string, received "${appId || ''}"`
      );
    }
    if (typeof apiKey !== 'string') {
      throw new AlgoliaError(
        `apiKey is required and should be a string, received ${apiKey}`
      );
    }
    if (typeof httpRequester !== 'function') {
      throw new AlgoliaError(
        `httpRequester is required and should be a function, received ${httpRequester}`
      );
    }

    initHostAndTimeouts({
      hosts,
      timeouts,
      appId,
    });

    this.appId = appId;
    this.apiKey = apiKey;
    this.requester = httpRequester;
    this.requestOptions = requestOptions;
    this.store = createMemoryStore();
    this.cache = cache;
  }

  setRequestOptions = (
    fn: RequestOptions => RequestOptions
  ): RequestOptions => {
    const oldOptions = this.requestOptions;
    const newOptions = fn(oldOptions);
    this.requestOptions = newOptions;
    return newOptions;
  };

  setRequesterOptions = (
    fn: RequesterOptions => RequesterOptions
  ): RequesterOptions => {
    const oldOptions = this.requesterOptions;
    const newOptions = fn(oldOptions);
    this.requesterOptions = newOptions;
    return newOptions;
  };

  saveInCache(key: string, data: Data): Data {
    if (this.cache) {
      return this.store.set(key, data);
    }

    return data;
  }

  getFromCache(key: string): ?Data {
    if (this.cache) {
      return this.store.get(key);
    }

    return undefined;
  }

  request = ({
    method,
    path,
    qs,
    body,
    requestOptions,
    // requesterOptions, // todo: implement
    requestType,
  }: RequestArguments): Promise<Result> => {
    try {
      const { appId, apiKey } = this;
      const { hostname, timeout, connectTimeout } = getParams({
        appId,
        requestType,
      });

      const pathname = path + stringify(qs);
      // todo: change to real URL
      const url = { hostname, pathname };

      const cacheKey = toCacheKey({
        method,
        url,
        path,
        qs,
        body,
        requestOptions,
        appId,
      });

      if (requestOptions && requestOptions.cache === true) {
        // check in cache
        const cachedResult = this.getFromCache(cacheKey);
        if (cachedResult !== undefined) {
          return Promise.resolve(cachedResult);
        }
      }

      return this.requester({
        body,
        method,
        url,
        timeout,
        connectTimeout,
        requestOptions, // but remove timeouts and cache here
        appId,
        apiKey,
      })
        .catch(err =>
          this.retryRequest(err, {
            method,
            path,
            qs,
            body,
            requestOptions,
            appId,
            apiKey,
            requestType,
          })
        )
        .then(res => this.saveInCache(cacheKey, res));
    } catch (e) {
      return Promise.reject(e);
    }
  };

  retryRequest = (
    err: AlgoliaRequesterError,
    requestArguments: RequestArguments
  ): Promise<Result> => {
    if (retryableErrors.indexOf(err.reason) > -1) {
      if (err.reason === 'timeout') {
        hostDidTimeout({
          appId: this.appId,
          requestType: requestArguments.requestType,
        });
      } else {
        hostDidFail({
          appId: this.appId,
          requestType: requestArguments.requestType,
        });
      }

      const res = this.request(requestArguments);

      return res;
    }

    return Promise.reject(
      // todo: use AlgoliaError and refactor?
      new AlgoliaRequesterError({
        message: `Request couldn't be retried, did you enter the correct credentials?

see: https://alg.li/client#unretryable-error`,
        reason: 'fatal',
        more: { error: err, arguments: requestArguments },
      })
    );
  };
}

export const createRequester: CreateRequester = function createRequester(args) {
  const _r = new Requester(args);
  const requester = _r.request;
  requester.setRequestOptions = _r.setRequestOptions;
  requester.options = _r.requestOptions;
  return requester;
};
