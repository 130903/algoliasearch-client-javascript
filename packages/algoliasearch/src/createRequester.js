import debug from './debug.js';
const log = debug('algolia:src/createRequester');

export default function createRequester({
  appId,
  apiKey,
  hosts,
  readProtocol,
  requester,
  timeouts
}) {
  return function request({
    method,
    path,
    pathParams,
    qs,
    body,
    forceReadHost
  }) {
    // const start = Date.now();

    let replaces = 0;

    // replace place holders like /%s/%s/ (most of the time, the indexName)
    path = path.replace(/%s/g, () => encodeURIComponent(pathParams[replaces++]));

    // add potential query string parameters to the path
    if (qs !== undefined) {
      path += `?${stringifyQueryStringObject(qs)}`;
    }

    log('%s %s', method, path);
    // console.log(body)

    return new Promise((resolve, reject) => {
      requester({
        body: JSON.stringify(body),
        headers: {
          'x-algolia-application-id': appId,
          'x-algolia-api-key': apiKey
        },
        hostname: hosts.write[0],
        method,
        onNetworkError: reject,
        onSuccess: res => {
          // console.log('it took ' + (Date.now() - start) + 'ms');
          resolve(JSON.parse(res.body));
        },
        port: 443,
        path,
        protocol: 'https:'
      });
    });

    // here should go the sw
    // to sending them in body
    // specific headers (per requester should be computed in requester itself)
  };
}

function stringifyQueryStringObject(qs) {
  return Object
    .keys(qs)
    .reduce((out, keyName, index) => {
      const value = qs[keyName];

      if (value === undefined || value === '') {
        return out;
      }

      return `${out}${index > 0 ? '&' : ''}${keyName}=${stringifyQueryStringValue(value)}`;
    }, '');
}

function stringifyQueryStringValue(value) {
  return encodeURIComponent(Array.isArray(value) ? JSON.stringify(value) : value);
}
