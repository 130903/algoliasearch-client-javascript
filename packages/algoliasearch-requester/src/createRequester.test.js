/* eslint prefer-promise-reject-errors: off */

import createRequester from './createRequester';

jest.useFakeTimers();

it('requires the right arguments', () => {
  const falseInvocations = [
    () =>
      createRequester({
        apiKey: '',
      }),
    () =>
      createRequester({
        appID: '',
      }),
    () =>
      createRequester({
        appID: '',
        apiKey: '',
        httpRequester: {},
      }),
  ];

  falseInvocations.map(invocation =>
    expect(invocation).toThrowErrorMatchingSnapshot()
  );

  expect(() =>
    createRequester({
      appID: '',
      apiKey: '',
      httpRequester: () => {},
    })
  ).not.toThrow();
});

it('first read request uses first host', () => {
  const httpRequester = jest.fn(() => Promise.resolve());
  const requester = createRequester({
    appID: 'the_read_app',
    apiKey: '',
    httpRequester,
  });

  requester({
    requestType: 'read',
  });

  const firstArgs = httpRequester.mock.calls[0];
  const { url: { hostname } } = firstArgs[0];
  expect(hostname).toEqual('the_read_app-dsn.algolia.net');
});

it('first write request uses first host', () => {
  const httpRequester = jest.fn(() => Promise.resolve());
  const requester = createRequester({
    appID: 'the_write_app',
    apiKey: '',
    httpRequester,
  });

  requester({
    requestType: 'write',
  });

  const firstArgs = httpRequester.mock.calls[0];
  const { url: { hostname } } = firstArgs[0];
  expect(hostname).toEqual('the_write_app.algolia.net');
});

it('uses a different host when the request needs to be retried', async () => {
  const httpRequester = jest.fn(
    () =>
      httpRequester.mock.calls.length === 1
        ? Promise.reject({ reason: 'network' })
        : Promise.resolve()
  );
  const requester = createRequester({
    appID: 'the_crazy_app',
    apiKey: '',
    httpRequester,
  });

  await requester({
    requestType: 'read',
  }); // retries

  const usedHosts = httpRequester.mock.calls.map(
    ([{ url: { hostname } }]) => hostname
  );

  expect(usedHosts).toMatchSnapshot();

  expect(usedHosts[0]).toEqual('the_crazy_app-dsn.algolia.net'); // first try
  expect(usedHosts[1]).toEqual('the_crazy_app-1.algolianet.com'); // second try
});

it('uses the "up" host on second request when first fails', async () => {
  const httpRequester = jest.fn(
    () =>
      httpRequester.mock.calls.length === 1
        ? Promise.reject({ reason: 'network' })
        : Promise.resolve()
  );
  const requester = createRequester({
    appID: 'the_crazy_app',
    apiKey: '',
    httpRequester,
  });

  await requester({
    requestType: 'read',
  }); // retries
  await requester({
    requestType: 'read',
  });

  const usedHosts = httpRequester.mock.calls.map(
    ([{ url: { hostname } }]) => hostname
  );

  expect(usedHosts).toMatchSnapshot();

  expect(usedHosts[0]).toEqual('the_crazy_app-dsn.algolia.net'); // first try
  expect(usedHosts[1]).toEqual('the_crazy_app-1.algolianet.com'); // first retry
  expect(usedHosts[2]).toEqual('the_crazy_app-1.algolianet.com'); // second request
});

it('resolves when the response is successful', () => {
  const httpRequester = jest.fn(() => Promise.resolve({}));
  const requester = createRequester({
    appID: 'the_successful_app',
    apiKey: '',
    httpRequester,
  });

  expect(
    requester({
      requestType: 'write',
    })
  ).resolves.toEqual({});
});

it("retries when there's a server error", async () => {
  const httpRequester = jest.fn(
    () =>
      httpRequester.mock.calls.length === 1
        ? Promise.reject({
            reason: 'server',
          })
        : Promise.resolve({ cool: 'turbo' })
  );
  const requester = createRequester({
    appID: 'the_app_app',
    apiKey: '',
    httpRequester,
  });

  // it eventually resolves
  await expect(
    requester({
      requestType: 'write',
    })
  ).resolves.toEqual({ cool: 'turbo' });

  // requester was called twice
  expect(httpRequester.mock.calls).toHaveLength(2);
});

it("retries when there's a network error", async () => {
  const httpRequester = jest.fn(
    () =>
      httpRequester.mock.calls.length === 1
        ? Promise.reject({
            reason: 'network',
          })
        : Promise.resolve({})
  );
  const requester = createRequester({
    appID: 'the_network_app',
    apiKey: '',
    httpRequester,
  });

  // it eventually resolves
  await expect(
    requester({
      requestType: 'write',
    })
  ).resolves.toEqual({});

  // requester was called twice
  expect(httpRequester.mock.calls).toHaveLength(2);
});

it("retries when there's a dns error", async () => {
  const httpRequester = jest.fn(
    () =>
      httpRequester.mock.calls.length === 1
        ? Promise.reject({
            reason: 'dns',
          })
        : Promise.resolve({})
  );
  const requester = createRequester({
    appID: 'the_dns_app',
    apiKey: '',
    httpRequester,
  });

  // it eventually resolves
  await expect(
    requester({
      requestType: 'write',
    })
  ).resolves.toEqual({});

  // requester was called twice
  expect(httpRequester.mock.calls.length).toBe(2);
});

it("retries when there's a timeout", async () => {
  const httpRequester = jest.fn(
    () =>
      httpRequester.mock.calls.length === 1
        ? Promise.reject({
            reason: 'timeout',
          })
        : Promise.resolve({ bingo: true })
  );
  const requester = createRequester({
    appID: 'the_retry_app',
    apiKey: '',
    httpRequester,
  });

  // it eventually resolves
  await expect(
    requester({
      requestType: 'write',
    })
  ).resolves.toEqual({ bingo: true });

  // requester was called twice
  expect(httpRequester.mock.calls.length).toBe(2);
});

it('second try after a timeout has increments the timeout (write)', async () => {
  const httpRequester = jest.fn(
    () =>
      httpRequester.mock.calls.length === 1
        ? Promise.reject({
            reason: 'timeout',
          })
        : Promise.resolve({})
  );
  const requester = createRequester({
    appID: 'the_fun_app',
    apiKey: '',
    httpRequester,
  });

  await requester({
    requestType: 'write',
  });

  const timeouts = httpRequester.mock.calls.map(([{ timeout }]) => timeout);

  expect(timeouts).toMatchSnapshot();

  expect(timeouts[1]).toBeGreaterThan(timeouts[0]);
});

it('second try after a timeout has increments the timeout (read)', async () => {
  const httpRequester = jest.fn(
    () =>
      httpRequester.mock.calls.length === 1
        ? Promise.reject({
            reason: 'timeout',
          })
        : Promise.resolve({})
  );
  const requester = createRequester({
    appID: 'the_fun_app',
    apiKey: '',
    httpRequester,
  });

  await requester({
    requestType: 'read',
  });

  const timeouts = httpRequester.mock.calls.map(([{ timeout }]) => timeout);

  expect(timeouts).toMatchSnapshot();

  expect(timeouts[1]).toBeGreaterThan(timeouts[0]);
});

it('rejects when all timeouts are reached', async () => {
  const httpRequester = jest.fn(
    () =>
      httpRequester.mock.calls.length <= 4
        ? Promise.reject({
            reason: 'timeout',
          })
        : Promise.resolve({})
  );
  const requester = createRequester({
    appID: 'the_timeout_app',
    apiKey: '',
    httpRequester,
  });

  await expect(
    requester({
      requestType: 'write',
    }) // eventually it will fail because it runs out of hosts
  ).rejects.toMatchSnapshot();

  await requester({ requestType: 'write' });

  const timeouts = httpRequester.mock.calls.map(([{ timeout }]) => timeout);
  const hosts = httpRequester.mock.calls.map(
    ([{ url: { hostname } }]) => hostname
  );
  expect(timeouts).toMatchSnapshot();

  const firstTimeout = timeouts[0];
  const lastTimeout = timeouts[timeouts.length - 1];

  expect(lastTimeout).toBe(firstTimeout);
});

it('rejects when all hosts are used', () => {
  const httpRequester = jest.fn(() =>
    Promise.reject({
      reason: 'network',
    })
  );
  const requester = createRequester({
    appID: 'the_host_app',
    apiKey: '',
    httpRequester,
  });

  // eventually it will fail because there are no more hosts
  expect(
    requester({
      requestType: 'write',
    })
  ).rejects.toMatchSnapshot('rejects when all hosts are used');
});

it.skip('uses the first host again after running out of hosts', async () => {
  const httpRequester = jest.fn(
    () =>
      httpRequester.mock.calls.length < 5 /* the request completely fails */
        ? Promise.reject({
            reason: 'network',
          })
        : Promise.resolve({})
  );
  const requester = createRequester({
    appID: 'the_other_host_app',
    apiKey: '',
    httpRequester,
  });

  // eventually it will fail because there are no more hosts
  await expect(
    requester({
      requestType: 'write',
    })
  ).rejects.toMatchSnapshot();

  await requester({
    requestType: 'write',
  }); // this request works

  const calls = httpRequester.mock.calls;
  const firstHost = calls[0][0].hostname;
  const lastHost = calls[calls.length - 1][0].hostname;

  expect(lastHost).toBe(firstHost);
});

it.skip('two instances of createRequester share the same host index', async () => {
  const httpRequester = jest.fn(
    () =>
      httpRequester.mock.calls.length === 1
        ? Promise.reject({ reason: 'network' })
        : Promise.resolve()
  );

  const firstRequester = createRequester({
    appID: 'the_same_app',
    apiKey: '',
    httpRequester,
  });
  const secondRequester = createRequester({
    appID: 'the_same_app',
    apiKey: '',
    httpRequester,
  });

  await firstRequester({
    requestType: 'read',
  }); // retries
  await secondRequester({
    requestType: 'read',
  });

  const usedHosts = httpRequester.mock.calls.map(
    ([{ url: { hostname } }]) => hostname
  );

  expect(usedHosts).toMatchSnapshot();

  expect(usedHosts[0]).toEqual('the_same_app-dsn.algolia.net'); // first try
  expect(usedHosts[1]).toEqual('the_same_app-1.algolianet.com'); // first retry
  expect(usedHosts[2]).toEqual('the_same_app-1.algolianet.com'); // second request
});
