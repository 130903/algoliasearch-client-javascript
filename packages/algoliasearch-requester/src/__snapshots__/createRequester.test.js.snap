// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`requires the right arguments 1`] = `"appId is required and should be a string, received \\"\\""`;

exports[`requires the right arguments 2`] = `"apiKey is required and should be a string, received undefined"`;

exports[`requires the right arguments 3`] = `"httpRequester is required and should be a function, received [object Object]"`;

exports[`uses a different host when the request needs to be retried 1`] = `
Array [
  Array [
    Object {
      "body": undefined,
      "method": undefined,
      "options": undefined,
      "timeout": 2000,
      "url": Object {
        "hostname": "the_crazy_app-dsn.algolia.net",
        "pathname": NaN,
      },
    },
  ],
]
`;
